python-threatexchange/threatexchange/hashing/pdq_faiss_matcher.py

```diff
@@ -101,6 +101,52 @@ def search(
             for i in range(len(query_vectors))
         ]
 
+    def search_with_distance(
+        self,
+        queries: t.Sequence[PDQ_HASH_TYPE],
+        threshhold: int,
+        return_as_ids: bool = False,
+    ):
+        """
+        Same as search however instead of returning only sequence of matches per query
+        it returns a mapping from query strings to a list of matched hashes (or ids) and distances
+        e.g.
+        result = {
+            "000000000000000000000000000000000000000000000000000000000000FFFF": [
+                ("00000000000000000000000000000000000000000000000000000000FFFFFFFF", 16.0)
+            ]
+        }
+        or if return_as_ids=True
+        result_with_as_ids = {
+            "000000000000000000000000000000000000000000000000000000000000FFFF": [
+                (12345678901, 16.0)
+            ]
+        }
+        """
+
+        query_vectors = [
+            numpy.frombuffer(binascii.unhexlify(q), dtype=numpy.uint8) for q in queries
+        ]
+        qs = numpy.array(query_vectors)
+        limits, similarities, I = self.faiss_index.range_search(qs, threshhold + 1)
+        if return_as_ids:
+            # for custom ids, we understood them initially as uint64 numbers and then coerced them internally to be signed
+            # int64s, so we need to reverse this before returning them back to the caller. For non custom ids, this will
+            # effectively return the same result
+            output_fn: t.Callable[[int], t.Any] = int64_to_uint64
+        else:
+            output_fn = self.hash_at
+
+        result = {}
+        for i in range(len(queries)):
```

Comment:
(ignorable, code golf): `for i, query in enumerate(queries):`

There's also stupid fancy things you can do zipping limits and queries together, but I can't tell what the below code is trying to do at a glance to advise you on a way to clean it up